### PR TITLE
add temporal helpers and models

### PR DESCRIPTION
**Pull request by YingtianDt**
This PR adds temporal model helpers and temporal models.

The editions are mainly constrained into the `model_helpers/activations/temporal` and corresponding tests.

### CODE REVIEW

File: brainscore_vision/model_helpers/activations/temporal/core/executor.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+from tqdm.auto import tqdm
+from collections import OrderedDict
+from typing import Any, Callable
+from ..inputs import Stimulus
+
+from brainscore_vision.model_helpers.utils import fullname
+from joblib import Parallel, delayed
+
+
+def pipeline(*funcs):
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/activations/temporal/core/executor.py:14]: comments pls

File: brainscore_vision/model_helpers/activations/temporal/core/executor.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    def _pipeline(x, *others):
+        for f in funcs:
+            x = f(x, *others)
+        return x
+    return _pipeline
+
+    
+class JoblibMapper:
+    def __init__(self, num_threads):
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/activations/temporal/core/executor.py:23]:
```suggestion
    def __init__(self, num_threads: int):
```

File: brainscore_vision/model_helpers/activations/temporal/core/executor.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    >>> executor = BatchExecutor(get_activations, preprocessing, batch_size=64, batch_grouper=lambda s: s.size)
+    >>> executor.register_before_hook(lambda v: v.set_size(180, 120))
+    >>> executor.register_after_hook(lambda val, l, s: val if l == "layer1" else val.mean(0))
+    >>> executor.add_stimuli(stimulus_list_1)
+    >>> executor.add_stimuli(stimulus_list_2)
+    >>> layer_activations = executor.execute(layers)  # here the actual batching and processing happen
+    """
+
+
+    def __init__(self, get_activations, preprocessing, batch_size, batch_padding,
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/activations/temporal/core/executor.py:72]: typehints pls

File: brainscore_vision/model_helpers/activations/temporal/core/extractor.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+
+        The hook should return a new stimulus_set which will be used in place of the previous one.
+        """
+
+        handle = HookHandle(self._stimulus_set_hooks)
+        self._stimulus_set_hooks[handle.id] = hook
+        return handle
+
+
+def change_dict(d, change_function, keep_name=False, multithread=False):
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/activations/temporal/core/extractor.py:150]: typehints, comments

File: brainscore_vision/model_helpers/activations/temporal/core/extractor.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+        return layer, values
+
+    results = map_fnc(apply_change, d.items())
+    results = OrderedDict(results)
+    if multithread:
+        pool.close()
+    return results
+
+
+def lstrip_local(path):
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/activations/temporal/core/extractor.py:169]: typehints, comments (I realize we never commented in the original code, but this is a good chance to improve)

File: brainscore_vision/model_helpers/activations/temporal/core/extractor.py [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+    parts = path.split(os.sep)
+    try:
+        start_index = parts.index('.brainio')
+    except ValueError:  # not in list -- perhaps custom directory
+        return path
+    path = os.sep.join(parts[start_index:])
+    return path
+
+
+def attach_stimulus_set_meta(assembly, stimulus_set):
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/activations/temporal/core/extractor.py:179]: typehints, comments

File: brainscore_vision/model_helpers/activations/temporal/core/extractor.py [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+    assert (np.array(assembly_paths) == np.array(stimulus_paths)).all()
+    assembly['stimulus_path'] = stimulus_set['stimulus_id'].values
+    assembly = assembly.rename({'stimulus_path': 'stimulus_id'})
+    for column in stimulus_set.columns:
+        assembly[column] = 'stimulus_id', stimulus_set[column].values
+    assembly = assembly.stack(presentation=('stimulus_id',))
+    return assembly
+
+
+class HookHandle:
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/activations/temporal/core/extractor.py:192]: brief comments

File: brainscore_vision/model_helpers/activations/temporal/core/extractor.py [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+
+    def disable(self):
+        self._saved_hook = self.remove()
+
+    def enable(self):
+        self.hook_dict[self.id] = self._saved_hook
+        self._saved_hook = None
+
+
+def flatten(layer_output, from_index=1, return_index=False):
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/activations/temporal/core/extractor.py:214]: typehints, brief comments

File: brainscore_vision/model_helpers/activations/temporal/core/inferencer/base.py [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+        self._stimulus_set_hooks = {}
+        self._batch_activations_hooks = {}
+        self._logger = logging.getLogger(fullname(self))
+
+        # register hooks
+        self._executor.register_after_hook(self._make_spatial_downsample_hook(max_spatial_size))
+        self._executor.register_after_hook(self._make_dtype_hook(dtype))
+
+    @property
+    def identifier(self):
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/activations/temporal/core/inferencer/base.py:104]:
```suggestion
    def identifier(self) -> str:
```

File: brainscore_vision/model_helpers/activations/temporal/core/inferencer/base.py [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+        to_add = [
+            f".dtype={self.dtype.__name__}",
+            f".vdeg={self.visual_degrees}",
+        ]
+        if self.max_spatial_size is not None:
+            to_add.append(f".max_s={self.max_spatial_size}")
+        to_add = "".join(to_add)
+        return f"{self.__class__.__name__}{to_add}"
+    
+    def set_visual_degrees(self, visual_degrees):
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/activations/temporal/core/inferencer/base.py:115]:
```suggestion
    def set_visual_degrees(self, visual_degrees: float):
```

File: brainscore_vision/model_helpers/activations/temporal/core/inferencer/base.py [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+        self.visual_degrees = visual_degrees
+
+    def __call__(self, paths, layers):
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/activations/temporal/core/inferencer/base.py:118]:
```suggestion
    def __call__(self, paths: List[Union[str, Path]], layers: List[str]):
```

File: brainscore_vision/model_helpers/activations/temporal/core/inferencer/video/base.py [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+        self.convert_to_video = convert_img_to_video
+
+    @property
+    def identifier(self):
+        id = f"{super().identifier}.{self.time_aligner.__name__}.fps={self.fps}"
+        if self.convert_to_video:
+            id += f".img_dur={self.img_duration}"
+        return id
+
+    def load_stimulus(self, path):
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/activations/temporal/core/inferencer/video/base.py:91]:
```suggestion
    def load_stimulus(self, path: Union[str, Path]]) -> Video:
```

File: brainscore_vision/model_helpers/activations/temporal/core/inferencer/video/base.py [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+
+    def _make_range(self, num, type="num_frames"):
+        if num is None:
+            return (1 if type=='num_frames' else 0, np.inf)
+        if isinstance(num, (tuple, list)):
+            return num
+        else:
+            return (num, num)
+
+    def _check_video(self, video):
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/activations/temporal/core/inferencer/video/base.py:124]:
```suggestion
    def _check_video(self, video: Video):
```

File: brainscore_vision/model_helpers/activations/temporal/core/inferencer/video/temporal_context/base.py [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+        - "fix": the length of the temporal context is determined by the specified "fixed_temporal_context".
+    
+    fixed_temporal_context: float
+        specify the fixed length of the temporal context, in ms. It will be used only if temporal_context_strategy is "fix".
+    
+    out_of_bound_strategy: str
+        specify how to handle the out-of-bound temporal context.
+        Options:
+        - "repeat": the out-of-bound temporal context will be repeated.
+        - TODO: "black": the out-of-bound temporal context will be zero-padded.
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/activations/temporal/core/inferencer/video/temporal_context/base.py:32]: what about gray padding? `128` is the usual background color people use

File: brainscore_vision/model_helpers/activations/temporal/core/inferencer/video/temporal_context/base.py [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+        ):
+        self.temporal_context_strategy = temporal_context_strategy
+        self.fixed_temporal_context = fixed_temporal_context
+        self.out_of_bound_strategy = out_of_bound_strategy
+        if self.temporal_context_strategy == "fix" and self.fixed_temporal_context is None:
+            raise ValueError("fixed_temporal_context must be specified if temporal_context_strategy is 'fix'.")
+        super().__init__(*args, **kwargs)
+
+    @property
+    def identifier(self):
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/activations/temporal/core/inferencer/video/temporal_context/base.py:50]:
```suggestion
    def identifier(self) -> str:
```

File: brainscore_vision/model_helpers/activations/temporal/core/inferencer/video/temporal_context/base.py [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+        lower, context = self._compute_temporal_context()
+        lower = round(lower, MS_ROUNDING_DIGITS)
+        context = round(context, MS_ROUNDING_DIGITS)
+        to_add = f".context={context}>{lower}.oob={self.out_of_bound_strategy}"
+        return f"{super().identifier}{to_add}"
+        
+    def load_stimulus(self, path):
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/activations/temporal/core/inferencer/video/temporal_context/base.py:57]:
```suggestion
    def load_stimulus(self, path: Union[str, Path]) -> Video:
```

File: brainscore_vision/model_helpers/activations/temporal/core/inferencer/video/temporal_context/base.py [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+        if self.convert_to_video and Stimulus.is_image_path(path):
+            video = Video.from_img_path(path, self.img_duration, self.fps)
+        else:
+            video = Video.from_path(path)
+        video = video.set_fps(self.fps)
+        # does no check here
+        return video
+
+    def _overlapped_range(self, s1, e1, s2, e2):
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/activations/temporal/core/inferencer/video/temporal_context/base.py:66]: I'm lost on what this method does. Please use descriptive variable names

File: brainscore_vision/model_helpers/activations/temporal/core/inferencer/video/temporal_context/causal.py [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+            **kwargs
+        ):
+        if "time_alignment" in kwargs:
+            if kwargs["time_alignment"] != "per_frame_aligned":
+                raise ValueError("CausalInferencer enforces time_alignment='per_frame_aligned'.")
+            else:
+                del kwargs["time_alignment"]
+        super().__init__(*args, **kwargs, time_alignment="per_frame_aligned")
+
+    def inference(self, stimuli, layers):
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/activations/temporal/core/inferencer/video/temporal_context/causal.py:33]: typehints, also return type `->`

File: brainscore_vision/model_helpers/activations/temporal/core/inferencer/video/time_aligner.py [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+        assembly = assembly.rename({"channel_temporal": "time_bin"})
+    else:
+        assembly = assembly.expand_dims("time_bin")
+    assembly = assembly.assign_coords({
+        "time_bin_start": ("time_bin", time_bin_starts),
+        "time_bin_end": ("time_bin", time_bin_ends)
+    })
+    return asm_type(assembly)
+
+def estimate_layer_fps(assembly, video):
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/activations/temporal/core/inferencer/video/time_aligner.py:19]: is the return type a float? typehints

File: brainscore_vision/model_helpers/activations/temporal/core/inferencer/video/time_aligner.py [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+    model_frame_interval = 1000 / fps
+    estimated_frame_interval = duration / num_t
+    estimated_multiplier = estimated_frame_interval / model_frame_interval
+    estimated_multiplier = int(round(estimated_multiplier))  # round to nearest integer
+    estimated_interval = model_frame_interval * estimated_multiplier
+    time_bin_starts = np.arange(0, num_t) * estimated_interval
+    time_bin_ends = time_bin_starts + estimated_interval
+    return _convert(assembly, time_bin_starts, time_bin_ends)
+
+def evenly_spaced(assembly, video):
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/activations/temporal/core/inferencer/video/time_aligner.py:34]: what does it return?

File: brainscore_vision/model_helpers/activations/temporal/core/inferencer/video/time_aligner.py [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+    # this function assumes that the activation of different time steps is evenly spaced
+    num_t = assembly.sizes['channel_temporal'] if "channel_temporal" in assembly.dims else 1
+    interval = video.duration / num_t
+    time_bin_starts = np.linspace(0, video.duration, num_t+1)[:-1]
+    time_bin_ends = time_bin_starts + interval
+    time_bin_ends[-1] = video.duration
+    return _convert(assembly, time_bin_starts, time_bin_ends)
+
+def per_frame_aligned(assembly, video):
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/activations/temporal/core/inferencer/video/time_aligner.py:43]: same question as the methods above

File: brainscore_vision/model_helpers/activations/temporal/core/inferencer/video/time_aligner.py [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+    # this function assumes that the activation of different time steps is aligned with the video frames
+    num_t = assembly.sizes['channel_temporal'] if "channel_temporal" in assembly.dims else 1
+    assert video.num_frames <= num_t
+    interval = 1000 / video.fps
+    time_bin_starts = np.arange(0, num_t) * interval
+    time_bin_ends = time_bin_starts + interval
+    return _convert(assembly, time_bin_starts, time_bin_ends)
+
+def ignore_time(assembly, video):
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/activations/temporal/core/inferencer/video/time_aligner.py:52]: does this return `Video`?

File: brainscore_vision/model_helpers/activations/temporal/inputs/base.py [hunk 23]
```diff
@@ -0,0 +1,14 @@
+class Stimulus:
+    def from_path(self, path):
+        raise NotImplementedError("Choose a concrete Stimulus type to use.")
+    
+    @staticmethod
+    def is_video_path(path):
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/activations/temporal/inputs/base.py:6]:
```suggestion
    def is_video_path(path: Union[str, Path]) -> bool:
```

File: brainscore_vision/model_helpers/activations/temporal/inputs/base.py [hunk 24]
```diff
@@ -0,0 +1,14 @@
...
+        extension = path.split('.')[-1]
+        return extension in ['mp4', 'avi', 'mov', 'flv', 'wmv', 'webm', 'mkv', 'gif']
+    
+    @staticmethod
+    def is_image_path(path):
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/activations/temporal/inputs/base.py:11]:
```suggestion
    def is_image_path(path: Union[str, Path]) -> bool
```

File: brainscore_vision/model_helpers/activations/temporal/inputs/image.py [hunk 25]
```diff
@@ -0,0 +1,50 @@
+import os
+import struct
+import numpy as np
+from PIL import Image as PILImage
+
+from .base import Stimulus
+
+
+def lazy_load(f):
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/activations/temporal/inputs/image.py:9]: what is `f`? use descriptive names and type hints

File: brainscore_vision/model_helpers/activations/temporal/inputs/video.py [hunk 26]
```diff
@@ -0,0 +1,147 @@
...
+    size = (width, height)
+    duration = length / fps * 1000
+    cap.release()
+    return fps, duration, size
+
+
+def get_image_stats(image_path):
+    img = PILImage.open(image_path)
+    size = img.size
+    img.close()
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/activations/temporal/inputs/video.py:27]:
```suggestion
    with PILImage.open(image_path) as img:
        return img.size
```
use context managers

File: brainscore_vision/model_helpers/activations/temporal/model/pytorch.py [hunk 27]
```diff
@@ -0,0 +1,103 @@
...
+        logger = logging.getLogger(fullname(self))
+        self._device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        logger.debug(f"Using device {self._device}")
+        self._model = model.to(self._device)
+        # preprocessing: input.Stimulus -> actual model inputs
+        self._preprocess = preprocessing
+        self._process_activation = default_process_activation if process_output is None else process_output
+        super().__init__(identifier, preprocessing, *args, **kwargs)
+
+    def forward(self, inputs):
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/activations/temporal/model/pytorch.py:38]: typehints

File: brainscore_vision/model_helpers/activations/temporal/model/pytorch.py [hunk 28]
```diff
@@ -0,0 +1,103 @@
...
+        import torch
+        tensor = torch.stack(inputs)
+        tensor = tensor.to(self._device)
+        return self._model(tensor)
+
+    def get_activations(self, inputs, layer_names):
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/activations/temporal/model/pytorch.py:44]: typehints

File: brainscore_vision/model_helpers/activations/temporal/utils.py [hunk 29]
```diff
@@ -0,0 +1,228 @@
...
+    inferencer = get_inferencer(any_model)
+    base_model = get_base_model(any_model)
+    for k, v in kwargs.items():
+        if v == 'same': kwargs[k] = getattr(inferencer, k)
+    base_model.build_extractor(new_inferencer_cls, **kwargs)
+
+
+## Others
+    
+def download_weight_file(url, folder=None):
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/activations/temporal/utils.py:199]: let's put this inside `brainscore_vision/model_helpers/s3.py`

File: brainscore_vision/model_helpers/brain_transformation/temporal.py [hunk 30]
```diff
@@ -7,10 +7,127 @@
...
+        return [iterable_to_list(a) for a in arr]
+    if isinstance(arr, np.ndarray):
+        if arr.dtype == object:
+            arr = arr.tolist()
+        return [iterable_to_list(a) for a in arr]
+    else:
+        return arr
+
+
+def time_align(source_time_bins: List[Tuple[int, int]], target_time_bins: List[Tuple[int, int]], mode: str = "portion"):
```
Comment (mschrimpf) [inline brainscore_vision/model_helpers/brain_transformation/temporal.py:22]: what's the output `->`?

File: brainscore_vision/models/temporal_model_SeLaVi/model.py [hunk 31]
```diff
@@ -0,0 +1,68 @@
...
+            state_dict[k] = v
+    model.load_state_dict(state_dict)
+
+    layer_activation_format = {
+        "base.stem": "CTHW",
+        **{f"base.layer{i}": "CTHW" for i in range(1, 5)},
+        # "base.fc": "C",  # no fc
+    }
+
+    return PytorchWrapper(identifier, model, transform_video, fps=30, layer_activation_format=layer_activation_format)
```
Comment (mschrimpf) [inline brainscore_vision/models/temporal_model_SeLaVi/model.py:68]: so `layer_activation_format` is really the list of layers right? I think we need to reflect that in the name. How about `layer_selection_and_format`?

File: brainscore_vision/models/temporal_model_VideoMAEv2/model.py [hunk 32]
```diff
@@ -0,0 +1,114 @@
+
+import numpy as np
+import torch
+from timm.models import create_model
+from torchvision import transforms
+
+# NOTE: Do not comment `import models`, it is used to register models
+from videomae_v2 import *  # noqa: F401
```
Comment (mschrimpf) [inline brainscore_vision/models/temporal_model_VideoMAEv2/model.py:8]: not a fan of `*` imports

File: brainscore_vision/models/temporal_model_mae_st/__init__.py [hunk 33]
```diff
@@ -0,0 +1,15 @@
...
+
+
+def commit_model(identifier):
+    activations_model=model.get_model(identifier)
+    layers=get_specified_layers(activations_model)
+    return ModelCommitment(identifier=identifier, activations_model=activations_model, layers=layers)
+
+
+model_registry["MAE-ST-L"] = lambda: commit_model("MAE-ST-L")
+# model_registry["MAE-ST-G"] = lambda: commit_model("MAE-ST-G")
```
Comment (mschrimpf) [inline brainscore_vision/models/temporal_model_mae_st/__init__.py:15]: dead code

**Review by mschrimpf — APPROVED** [review]
Looks good with the following caveats:
* please add typehints where possible, for the variables (e.g. `: float`) as well as the return type (e.g. `-> Video`, except where there is no return then we don't have to write out the `-> None`).
* please comment on especially the high-level functions and potentially more difficult functions (e.g. many lines of code) and what they do
* add a `test.py` file to the models. This can be as simple as making sure they load, or making sure they reproduce some benchmark score
* I'm also having some trouble understanding all the unit tests, maybe you can add a brief comment on at least the file/module level to explain what we're testing for

I also think it might be easier to separate this into multiple PRs: one for the `model_helpers` addition with just one model plugin being added as an example. And then more PRs for all the other model plugin additions. We can also try all of them at the same time if you prefer, lmk.

File: brainscore_vision/model_helpers/activations/temporal/utils.py (hunk 29 above)
Comment (YingtianDt) [reply brainscore_vision/model_helpers/activations/temporal/utils.py:199]: This code has nothing to do with s3 though. Are you sure?

**Comment by mschrimpf** [comment]
re-trigger jenkins after node error

File: brainscore_vision/model_interface.py [hunk 34]
```diff
@@ -169,6 +169,7 @@ class RecordingTarget:
         V2 = 'V2'
         V4 = 'V4'
         IT = 'IT'
+        whole_brain = "whole_brain"
```
Comment (mschrimpf) [inline brainscore_vision/model_interface.py:172]: please undo this change. We can discuss in another PR

**Comment by mschrimpf** [comment]
all tests pass 
image
including for the model plugin
image
